> Hi @sayboras , do you need help or feedback on this issue/PR? It is currently marked as a release blocker for v1.8 so we will want to determine whether it's vital to merge for v1.8 and if so, help to get it merged.

@joestringer sure, if you can help on this. I added lock for SpanStat struct, however, was having below copylocks issue.

To resolve this, I need to use pointer (e.g. lock *lock.RWMutex), but then the drawback is to have constructor to initialize values. This must be done for all instances of SpanStat struct. 

Looking forward to hearing your input. Thanks

<details>
<summary> copylocks issue </summary>

```shell scripts
$ golangci-lint run ./... --disable-all -E govet 
daemon/cmd/fqdn.go:408:164: copylocks: notifyOnDNSMsg passes lock by value: github.com/cilium/cilium/pkg/fqdn/dnsproxy.ProxyRequestContext contains github.com/cilium/cilium/pkg/spanstat.SpanStat contains github.com/cilium/cilium/pkg/lock.RWMutex (govet)
func (d *Daemon) notifyOnDNSMsg(lookupTime time.Time, ep *endpoint.Endpoint, epIPPort string, serverAddr string, msg *dns.Msg, protocol string, allowed bool, stat dnsproxy.ProxyRequestContext) error {
                                                                                                                                                                   ^
pkg/endpoint/policy.go:215:33: copylocks: literal copies lock value from stats.waitingForIdentityCache: github.com/cilium/cilium/pkg/spanstat.SpanStat contains github.com/cilium/cilium/pkg/lock.RWMutex (govet)
                "waitingForIdentityCache":    stats.waitingForIdentityCache,
                                              ^
pkg/endpoint/policy.go:216:33: copylocks: literal copies lock value from stats.waitingForPolicyRepository: github.com/cilium/cilium/pkg/spanstat.SpanStat contains github.com/cilium/cilium/pkg/lock.RWMutex (govet)
                "waitingForPolicyRepository": stats.waitingForPolicyRepository,
                                              ^
pkg/endpoint/policy.go:217:33: copylocks: literal copies lock value from stats.policyCalculation: github.com/cilium/cilium/pkg/spanstat.SpanStat contains github.com/cilium/cilium/pkg/lock.RWMutex (govet)
                "policyCalculation":          stats.policyCalculation,
                                              ^
pkg/fqdn/dnsproxy/proxy.go:373:77: copylocks: call of p.NotifyOnDNSMsg copies lock value: github.com/cilium/cilium/pkg/fqdn/dnsproxy.ProxyRequestContext contains github.com/cilium/cilium/pkg/spanstat.SpanStat contains github.com/cilium/cilium/pkg/lock.RWMutex (govet)
                p.NotifyOnDNSMsg(time.Now(), nil, epIPPort, "", request, protocol, false, stat)
                                                                                          ^
pkg/fqdn/dnsproxy/proxy.go:382:77: copylocks: call of p.NotifyOnDNSMsg copies lock value: github.com/cilium/cilium/pkg/fqdn/dnsproxy.ProxyRequestContext contains github.com/cilium/cilium/pkg/spanstat.SpanStat contains github.com/cilium/cilium/pkg/lock.RWMutex (govet)
                p.NotifyOnDNSMsg(time.Now(), nil, epIPPort, "", request, protocol, false, stat)
                                                                                          ^
pkg/fqdn/dnsproxy/proxy.go:394:90: copylocks: call of p.NotifyOnDNSMsg copies lock value: github.com/cilium/cilium/pkg/fqdn/dnsproxy.ProxyRequestContext contains github.com/cilium/cilium/pkg/spanstat.SpanStat contains github.com/cilium/cilium/pkg/lock.RWMutex (govet)
                p.NotifyOnDNSMsg(time.Now(), ep, epIPPort, targetServerAddr, request, protocol, false, stat)
```

</details>

--- a/daemon/cmd/fqdn.go
+++ b/daemon/cmd/fqdn.go
@@ -405,7 +405,7 @@ func (d *Daemon) lookupEPByIP(endpointIP net.IP) (endpoint *endpoint.Endpoint, e
 //   can lookup the endpoint related to it
 // epIPPort and serverAddr should match the original request, where epAddr is
 // the source for egress (the only case current).
-func (d *Daemon) notifyOnDNSMsg(lookupTime time.Time, ep *endpoint.Endpoint, epIPPort string, serverAddr string, msg *dns.Msg, protocol string, allowed bool, stat dnsproxy.ProxyRequestContext) error {
+func (d *Daemon) notifyOnDNSMsg(lookupTime time.Time, ep *endpoint.Endpoint, epIPPort string, serverAddr string, msg *dns.Msg, protocol string, allowed bool, stat *dnsproxy.ProxyRequestContext) error {
 	var protoID = u8proto.ProtoIDs[strings.ToLower(protocol)]
 	var verdict accesslog.FlowVerdict
 	var reason string

--- a/pkg/endpoint/policy.go
+++ b/pkg/endpoint/policy.go
@@ -212,9 +212,9 @@ func (e *Endpoint) updatePolicyRegenerationStatistics(stats *policyRegenerationS
 	stats.SendMetrics()
 
 	fields := logrus.Fields{
-		"waitingForIdentityCache":    stats.waitingForIdentityCache,
-		"waitingForPolicyRepository": stats.waitingForPolicyRepository,
-		"policyCalculation":          stats.policyCalculation,
+		"waitingForIdentityCache":    &stats.waitingForIdentityCache,
+		"waitingForPolicyRepository": &stats.waitingForPolicyRepository,
+		"policyCalculation":          &stats.policyCalculation,
 		"forcedRegeneration":         forceRegeneration,
 	}
 	scopedLog := e.getLogger().WithFields(fields)

--- a/pkg/fqdn/dnsproxy/proxy.go
+++ b/pkg/fqdn/dnsproxy/proxy.go
@@ -205,7 +205,7 @@ type LookupSecIDByIPFunc func(ip net.IP) (secID ipcache.Identity, exists bool)
 
 // NotifyOnDNSMsgFunc handles propagating DNS response data
 // See DNSProxy.LookupEndpointIDByIP for usage.
-type NotifyOnDNSMsgFunc func(lookupTime time.Time, ep *endpoint.Endpoint, epIPPort string, serverAddr string, msg *dns.Msg, protocol string, allowed bool, stat ProxyRequestContext) error
+type NotifyOnDNSMsgFunc func(lookupTime time.Time, ep *endpoint.Endpoint, epIPPort string, serverAddr string, msg *dns.Msg, protocol string, allowed bool, stat *ProxyRequestContext) error
 
 // ProxyRequestContext proxy dns request context struct to send in the callback
 type ProxyRequestContext struct {
@@ -370,7 +370,7 @@ func (p *DNSProxy) ServeDNS(w dns.ResponseWriter, request *dns.Msg) {
 		scopedLog.WithError(err).Error("cannot extract endpoint IP from DNS request")
 		stat.Err = fmt.Errorf("Cannot extract endpoint IP from DNS request: %s", err)
 		stat.ProcessingTime.End(false)
-		p.NotifyOnDNSMsg(time.Now(), nil, epIPPort, "", request, protocol, false, stat)
+		p.NotifyOnDNSMsg(time.Now(), nil, epIPPort, "", request, protocol, false, &stat)
 		p.sendRefused(scopedLog, w, request)
 		return
 	}
@@ -379,7 +379,7 @@ func (p *DNSProxy) ServeDNS(w dns.ResponseWriter, request *dns.Msg) {
 		scopedLog.WithError(err).Error("cannot extract endpoint ID from DNS request")
 		stat.Err = fmt.Errorf("Cannot extract endpoint ID from DNS request: %s", err)
 		stat.ProcessingTime.End(false)
-		p.NotifyOnDNSMsg(time.Now(), nil, epIPPort, "", request, protocol, false, stat)
+		p.NotifyOnDNSMsg(time.Now(), nil, epIPPort, "", request, protocol, false, &stat)
 		p.sendRefused(scopedLog, w, request)
 		return
 	}
@@ -391,7 +391,7 @@ func (p *DNSProxy) ServeDNS(w dns.ResponseWriter, request *dns.Msg) {
 		log.WithError(err).Error("cannot extract destination IP:port from DNS request")
 		stat.Err = fmt.Errorf("Cannot extract destination IP:port from DNS request: %s", err)
 		stat.ProcessingTime.End(false)
-		p.NotifyOnDNSMsg(time.Now(), ep, epIPPort, targetServerAddr, request, protocol, false, stat)
+		p.NotifyOnDNSMsg(time.Now(), ep, epIPPort, targetServerAddr, request, protocol, false, &stat)
 		p.sendRefused(scopedLog, w, request)
 		return
 	}
@@ -401,7 +401,7 @@ func (p *DNSProxy) ServeDNS(w dns.ResponseWriter, request *dns.Msg) {
 		scopedLog.WithField("server", targetServerAddr).Debug("cannot find server ip in ipcache")
 		stat.Err = fmt.Errorf("cannot find security identity for server ip=%s", targetServerIP)
 		stat.ProcessingTime.End(false)
-		p.NotifyOnDNSMsg(time.Now(), ep, epIPPort, targetServerAddr, request, protocol, false, stat)
+		p.NotifyOnDNSMsg(time.Now(), ep, epIPPort, targetServerAddr, request, protocol, false, &stat)
 		p.sendRefused(scopedLog, w, request)
 		return
 	}
@@ -418,7 +418,7 @@ func (p *DNSProxy) ServeDNS(w dns.ResponseWriter, request *dns.Msg) {
 		scopedLog.WithError(err).Error("Rejecting DNS query from endpoint due to error")
 		stat.Err = fmt.Errorf("Rejecting DNS query from endpoint due to error: %s", err)
 		stat.ProcessingTime.End(false)
-		p.NotifyOnDNSMsg(time.Now(), ep, epIPPort, targetServerAddr, request, protocol, false, stat)
+		p.NotifyOnDNSMsg(time.Now(), ep, epIPPort, targetServerAddr, request, protocol, false, &stat)
 		p.sendRefused(scopedLog, w, request)
 		return
 
@@ -426,12 +426,12 @@ func (p *DNSProxy) ServeDNS(w dns.ResponseWriter, request *dns.Msg) {
 		scopedLog.Debug("Rejecting DNS query from endpoint due to policy")
 		stat.Err = p.sendRefused(scopedLog, w, request)
 		stat.ProcessingTime.End(true)
-		p.NotifyOnDNSMsg(time.Now(), ep, epIPPort, targetServerAddr, request, protocol, false, stat)
+		p.NotifyOnDNSMsg(time.Now(), ep, epIPPort, targetServerAddr, request, protocol, false, &stat)
 		return
 	}
 
 	scopedLog.Debug("Forwarding DNS request for a name that is allowed")
-	p.NotifyOnDNSMsg(time.Now(), ep, epIPPort, targetServerAddr, request, protocol, true, stat)
+	p.NotifyOnDNSMsg(time.Now(), ep, epIPPort, targetServerAddr, request, protocol, true, &stat)
 
 	// Keep the same L4 protocol. This handles DNS re-requests over TCP, for
 	// requests that were too large for UDP.
@@ -445,7 +445,7 @@ func (p *DNSProxy) ServeDNS(w dns.ResponseWriter, request *dns.Msg) {
 		scopedLog.Error("Cannot parse DNS proxy client network to select forward client")
 		stat.Err = fmt.Errorf("Cannot parse DNS proxy client network to select forward client: %s", err)
 		stat.ProcessingTime.End(false)
-		p.NotifyOnDNSMsg(time.Now(), ep, epIPPort, targetServerAddr, request, protocol, false, stat)
+		p.NotifyOnDNSMsg(time.Now(), ep, epIPPort, targetServerAddr, request, protocol, false, &stat)
 		p.sendRefused(scopedLog, w, request)
 		return
 	}
@@ -464,7 +464,7 @@ func (p *DNSProxy) ServeDNS(w dns.ResponseWriter, request *dns.Msg) {
 			p.sendRefused(scopedLog, w, request)
 			stat.Err = fmt.Errorf("Cannot forward proxied DNS lookup: %s", err)
 		}
-		p.NotifyOnDNSMsg(time.Now(), ep, epIPPort, targetServerAddr, request, protocol, false, stat)
+		p.NotifyOnDNSMsg(time.Now(), ep, epIPPort, targetServerAddr, request, protocol, false, &stat)
 		return
 	}
 
@@ -472,7 +472,7 @@ func (p *DNSProxy) ServeDNS(w dns.ResponseWriter, request *dns.Msg) {
 	stat.Success = true
 
 	scopedLog.Debug("Notifying with DNS response to original DNS query")
-	p.NotifyOnDNSMsg(time.Now(), ep, epIPPort, targetServerAddr, response, protocol, true, stat)
+	p.NotifyOnDNSMsg(time.Now(), ep, epIPPort, targetServerAddr, response, protocol, true, &stat)
 
 	scopedLog.Debug("Responding to original DNS query")
 	// restore the ID to the one in the initial request so it matches what the requester expects.
@@ -482,7 +482,7 @@ func (p *DNSProxy) ServeDNS(w dns.ResponseWriter, request *dns.Msg) {
 	if err != nil {
 		scopedLog.WithError(err).Error("Cannot forward proxied DNS response")
 		stat.Err = fmt.Errorf("Cannot forward proxied DNS response: %s", err)
-		p.NotifyOnDNSMsg(time.Now(), ep, epIPPort, targetServerAddr, response, protocol, true, stat)
+		p.NotifyOnDNSMsg(time.Now(), ep, epIPPort, targetServerAddr, response, protocol, true, &stat)
 	}
 }
 

--- a/pkg/fqdn/dnsproxy/proxy_test.go
+++ b/pkg/fqdn/dnsproxy/proxy_test.go
@@ -199,7 +199,7 @@ func (s *DNSProxyTestSuite) SetUpSuite(c *C) {
 			}
 		},
 		// NotifyOnDNSMsg
-		func(lookupTime time.Time, ep *endpoint.Endpoint, epIPPort string, dstAddr string, msg *dns.Msg, protocol string, allowed bool, stat ProxyRequestContext) error {
+		func(lookupTime time.Time, ep *endpoint.Endpoint, epIPPort string, dstAddr string, msg *dns.Msg, protocol string, allowed bool, stat *ProxyRequestContext) error {
 			return nil
 		})
 	c.Assert(err, IsNil, Commentf("error starting DNS Proxy"))

--- a/pkg/spanstat/spanstat.go
+++ b/pkg/spanstat/spanstat.go
@@ -17,6 +17,7 @@ package spanstat
 import (
 	"time"
 
+	"github.com/cilium/cilium/pkg/lock"
 	"github.com/cilium/cilium/pkg/logging"
 	"github.com/cilium/cilium/pkg/logging/logfields"
 	"github.com/cilium/cilium/pkg/safetime"
@@ -30,6 +31,7 @@ var (
 // SpanStat measures the total duration of all time spent in between Start()
 // and Stop() calls.
 type SpanStat struct {
+	mutex           lock.RWMutex
 	spanStart       time.Time
 	successDuration time.Duration
 	failureDuration time.Duration
@@ -43,19 +45,30 @@ func Start() *SpanStat {
 
 // Start starts a new span
 func (s *SpanStat) Start() *SpanStat {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
 	s.spanStart = time.Now()
 	return s
 }
 
 // EndError calls End() based on the value of err
 func (s *SpanStat) EndError(err error) *SpanStat {
-	return s.End(err == nil)
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	return s.end(err == nil)
 }
 
 // End ends the current span and adds the measured duration to the total
 // cumulated duration, and to the success or failure cumulated duration
 // depending on the given success flag
 func (s *SpanStat) End(success bool) *SpanStat {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	return s.end(success)
+}
+
+// must be called with Lock() held
+func (s *SpanStat) end(success bool) *SpanStat {
 	if !s.spanStart.IsZero() {
 		d, _ := safetime.TimeSinceSafe(s.spanStart, log)
 		if success {
@@ -71,21 +84,29 @@ func (s *SpanStat) End(success bool) *SpanStat {
 // Total returns the total duration of all spans measured, including both
 // successes and failures
 func (s *SpanStat) Total() time.Duration {
+	s.mutex.RLock()
+	defer s.mutex.RUnlock()
 	return s.successDuration + s.failureDuration
 }
 
 // SuccessTotal returns the total duration of all successful spans measured
 func (s *SpanStat) SuccessTotal() time.Duration {
+	s.mutex.RLock()
+	defer s.mutex.RUnlock()
 	return s.successDuration
 }
 
 // FailureTotal returns the total duration of all unsuccessful spans measured
 func (s *SpanStat) FailureTotal() time.Duration {
+	s.mutex.RLock()
+	defer s.mutex.RUnlock()
 	return s.failureDuration
 }
 
 // Reset rests the duration measurements
 func (s *SpanStat) Reset() {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
 	s.successDuration = 0
 	s.failureDuration = 0
 }
@@ -93,9 +114,12 @@ func (s *SpanStat) Reset() {
 // Seconds returns the number of seconds represents by the spanstat. If a span
 // is still open, it is closed first.
 func (s *SpanStat) Seconds() float64 {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
 	if !s.spanStart.IsZero() {
-		s.End(true)
+		s.end(true)
 	}
 
-	return s.Total().Seconds()
+	total := s.successDuration + s.failureDuration
+	return total.Seconds()
 }

--- a/pkg/spanstat/spanstat_test.go
+++ b/pkg/spanstat/spanstat_test.go
@@ -17,9 +17,12 @@
 package spanstat
 
 import (
+	"fmt"
+	"sync"
 	"testing"
 	"time"
 
+	"github.com/stretchr/testify/assert"
 	. "gopkg.in/check.v1"
 )
 
@@ -89,4 +92,101 @@ func (s *SpanStatTestSuite) TestSpanStat(c *C) {
 func (s *SpanStatTestSuite) TestSpanStatSeconds(c *C) {
 	span1 := Start()
 	c.Assert(span1.Seconds(), Not(Equals), float64(0))
+}
+
+func (s *SpanStatTestSuite) TestSpanStatSecondsRaceCondition(c *C) {
+	span1 := Start()
+	var wg sync.WaitGroup
+
+	for i := 0; i < 10; i++ {
+		wg.Add(1)
+		go func(span *SpanStat) {
+			c.Assert(span1.Seconds(), Not(Equals), float64(0))
+		}(span1)
+	}
+	wg.Done()
+}
+
+func TestSpanStatRaceCondition(t *testing.T) {
+	type fields struct {
+		runFunc func(span *SpanStat) float64
+	}
+	tests := []struct {
+		name   string
+		fields fields
+	}{
+		{
+			name: "End function",
+			fields: fields{
+				runFunc: func(span *SpanStat) float64 {
+					return span.End(true).Seconds()
+				},
+			},
+		},
+		{
+			name: "EndError function",
+			fields: fields{
+				runFunc: func(span *SpanStat) float64 {
+					return span.EndError(fmt.Errorf("dummy error")).Seconds()
+				},
+			},
+		},
+		{
+			name: "Seconds function",
+			fields: fields{
+				runFunc: func(span *SpanStat) float64 {
+					return span.Seconds()
+				},
+			},
+		},
+		{
+			name: "Total function",
+			fields: fields{
+				runFunc: func(span *SpanStat) float64 {
+					return span.Total().Seconds() + 1
+				},
+			},
+		},
+		{
+			name: "FailureTotal function",
+			fields: fields{
+				runFunc: func(span *SpanStat) float64 {
+					return span.FailureTotal().Seconds() + 1
+				},
+			},
+		},
+		{
+			name: "SuccessTotal function",
+			fields: fields{
+				runFunc: func(span *SpanStat) float64 {
+					return span.SuccessTotal().Seconds() + 1
+				},
+			},
+		},
+		{
+			name: "Reset function",
+			fields: fields{
+				runFunc: func(span *SpanStat) float64 {
+					span.Reset()
+					return 1
+				},
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			span := Start()
+			var wg sync.WaitGroup
+
+			for i := 0; i < 5; i++ {
+				wg.Add(1)
+				go func(span *SpanStat) {
+					assert.NotEqual(t, tt.fields.runFunc(span), float64(0))
+				}(span)
+			}
+			wg.Done()
+		})
+	}
+
 }

